File: app/src/main/java/com/cellblock70/popularmovies/data/database/MovieDao.java
package com.cellblock70.popularmovies.data.database;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Transaction;

import java.util.List;

@Dao
public interface MovieDao {

    @Transaction
    @Query("SELECT * FROM details WHERE movie_id IN (SELECT * FROM favorite)")
    List<CompleteMovie> getFavorites();

    @Delete
    void delete(Favorite favorite);
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insert(Favorite favorite);

    // TODO delete unused movies
    @Query("DELETE FROM details WHERE movie_id NOT IN (SELECT movie_id FROM favorite)")
    void deleteAllExceptFavorites();

    @Transaction
    @Query("SELECT * FROM details where movie_id = (:movieId)")
    LiveData<CompleteMovie> getMovieWithTrailersAndReviews(Integer movieId);

    @Transaction
    @Query("SELECT * FROM details where movie_id = (:movieId)")
    CompleteMovie getCompleteMovie(Integer movieId);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertAll(Movie...movieDetails);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertAll(MovieTrailer...trailers);

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertAll(MovieReview...reviews);
}

File: app/src/main/java/com/cellblock70/popularmovies/data/database/MovieDao.kt
package com.cellblock70.popularmovies.data.database

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import kotlinx.coroutines.flow.Flow

@Dao
interface MovieDao {

    @Query("SELECT * FROM favorite")
    suspend fun getFavoriteList() : List<Favorite>

    @Query("SELECT EXISTS(SELECT * FROM favorite WHERE movie_id = :movieId)")
    fun isFavorite(movieId: Int): Boolean

    @Delete
    suspend fun deleteFavorite(favorite: Favorite)

    @Query("DELETE FROM details")
    suspend fun deleteAllMovies()

    @Query("SELECT * FROM details WHERE movie_id = (:movieId)")
    fun getMovie(movieId: Int): Movie

    @Query("SELECT * FROM details")
    fun getMovieList(): Flow<List<Movie>>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insert(favorite: Favorite)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertAll(vararg movieDetails: Movie)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertAll(vararg trailers: MovieTrailer)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertAll(vararg reviews: MovieReview)

    @Query("SELECT * FROM reviews WHERE movie_id = (:movieId)")
    fun getReviews(movieId: Int): List<MovieReview>

    @Query("SELECT * FROM trailers WHERE movie_id = (:movieId)")
    fun getTrailers(movieId: Int): List<MovieTrailer>
}